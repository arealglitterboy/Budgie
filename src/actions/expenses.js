import {v4 as uuid} from 'uuid';

export const addExpense = (
    {
        description = '',
        note = '',
        date = 0,
        currency = 'â‚¬',
        amount = 0
    } = {}
    ) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        date,
        currency,
        amount
    }
});

export const removeExpense = ({ id }) => ({ type: 'REMOVE_EXPENSE', id });

export const editExpense = (id, updates) => ({ type: 'EDIT_EXPENSE', id, updates });