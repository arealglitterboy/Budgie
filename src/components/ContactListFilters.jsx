'use strict';
import React from 'react';

import { connect } from 'react-redux';
import { setCategories, setFilterTerm, setSortBy } from '../actions/contactsFilters.action';

import SearchInput from './filter/SearchInput';
import CategoriesSelect from './filter/CategoriesSelect';
import SortSelect from './filter/SortSelect';

const option = (value, label) => ({ value, label });

const options = [
    option('byNewest', 'By Newest'),
    option('byOldest', 'By Oldest'),
    option('byNameDescending', 'Name Descending'),
    option('byNameAscending', 'Name Ascending')
];

export const ContactListFilters = (props) => {
    const clearFilters = () => {
        props.setSortBy();
        props.setFilterTerm();
        props.setCategories();
    };

    return (
        <header className="contacts-filters">
            <SearchInput 
                search={props.filters.term}
                setSearch={props.setFilterTerm}
            />
            <SortSelect 
                options={options}
                sortBy={props.filters.sortBy}
                setSortBy={props.setSortBy}
            />
            <CategoriesSelect 
                categories={props.filters.categories}
                setCategories={props.setCategories}
            />
            <button className="contacts-filters__clear" onClick={clearFilters}>clear</button>
        </header>
    );
}


const mapDispatchToProps = (dispatch) => ({
    setFilterTerm: (term) => dispatch(setFilterTerm(term)),
    setSortBy: (sortBy) => dispatch(setSortBy(sortBy)),
    setCategories: (categories) => dispatch(setCategories(categories))
});

const mapStateToProps = connect(({ contactsFilters }) => ({ filters: contactsFilters }), mapDispatchToProps);

export default mapStateToProps(ContactListFilters);