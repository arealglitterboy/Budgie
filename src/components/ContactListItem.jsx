import React from 'react'

const imgOnError = (e) => {
    e.target.src = `data:image/svg+xml;charset=UTF-8,%3csvg xmlns='http://www.w3.org/2000/svg' height='48px' viewBox='0 0 24 24' width='48px' %3e%3cpath d='M0 0h24v24H0V0z' fill='none' /%3e%3cpath d='M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z' /%3e%3c/svg%3e`;
    e.target.classList.add('contact__icon__image--default');
    e.target.onError = () => {};
};

const buttonOnFocus = (e) => {
    console.log();
    e.target.nextSibling.classList.add('contact__categories__list--active')
};

const buttonOnBlur = (e) => {
    e.target.nextSibling.classList.remove('contact__categories__list--active')
}

export default (props) => (
    <article className="contact" id={props.id}>
        <div className="contact__icon">
            <img src={props.icon || ''} alt="" className="contact__icon__image" onError={imgOnError} />
        </div>
        <h5 className="contact__name">{props.name}</h5>
        <p className="contact__date">
            <time dateTime={props.date}>{props.date.toLocaleDateString()}</time>
        </p>
        <section className="contact__categories">
            <button className="contact__categories__button" onFocus={buttonOnFocus} onBlur={buttonOnBlur}>view</button>
            <ul className="contact__categories__list">
                {props.categories.map((val, i) => <li key={i} className="contact__categories__list__item">{val}</li>)}
            </ul>
        </section>
    </article>
)