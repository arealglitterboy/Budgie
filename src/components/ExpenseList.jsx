import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses.selector';

const ExpenseList = ({ expenses }) => (
    <section className="expense-list">
        {expenses.map((e) => (<ExpenseListItem { ...e } key={e.id} />))}
    </section>
);

const mapStateToProps = connect(({ expenses, filters }) => ({
    expenses: selectExpenses(expenses, filters)
}));

export default mapStateToProps(ExpenseList);