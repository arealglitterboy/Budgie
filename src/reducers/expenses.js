const defaultState = [];

export default (state = defaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter((e) => (e.id !== action.id));
        case 'EDIT_EXPENSE': // ? Map each value in the array into a new array, if the id matches the selected id map it with its new values
            return state.map((e) => ((e.id === action.id) ? {...e, ...action.updates} : {...e}));
        default:
            return state;
    }
};