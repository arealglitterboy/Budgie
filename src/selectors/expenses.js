import moment from "moment";

const includesIgnoreCase = (key, searching) => (!key || (searching && searching.toLowerCase().includes(key.toLowerCase())));

/**
 * 
 * @param {string} search 
 * @returns {function}
 */
const createIncludes = (search) => {
    const regExp = new RegExp(search, 'gi')
    return (str) => str.search(regExp) >= 0;
};

function filterExpense({ description, note, date }, term, startDate, endDate) {
    const includes = createIncludes(term);
    const dateMoment = moment(date);

    const containsTerm = includes(description) || includes(note); // ? If either the description or the note contains the search term.
    // const containsTerm = includesIgnoreCase(term, description) || includesIgnoreCase(term, note); // ? If either the description or the note contains the search term.
    const inDate = (!startDate || startDate.isSameOrBefore(dateMoment, 'day')) && (!endDate || endDate.isSameOrAfter(dateMoment, 'day')); // ? If the start/end date is defined, and the given date is greater/less than it.
    return containsTerm && inDate;
}

export default (expenses, { term, startDate, endDate, sortFunction }) => {
    return expenses.filter((e) => (filterExpense(e, term, startDate, endDate))).sort(sortFunction);
};