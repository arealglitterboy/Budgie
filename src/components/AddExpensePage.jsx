import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses.action';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <main>
                <h2>Create Expense</h2>
                <ExpenseForm onSubmit={this.onSubmit} />
            </main>
        );    
    }
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
});

const mapStateToProps = connect(undefined, mapDispatchToProps);

export default mapStateToProps(AddExpensePage);