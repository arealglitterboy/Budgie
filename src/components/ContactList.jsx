import React from 'react'
import { connect } from 'react-redux'

import ContactListItem from './ContactListItem';
import selectContacts from '../selectors/contacts.selector';

export const ContactList = ({ contacts=[] }) => (
    <section className="contacts-list">
        {
            (contacts.length === 0) 
            ? <p className="contacts__empty-message">Start by adding</p>
            : contacts.map((contact) => (<ContactListItem {...contact} id={contact.id} name={contact.name} key={contact.id} />))
        }
    </section>
);

const mapStateToProps = connect(({ contacts, contactsFilters }) => ({
    contacts: selectContacts(contacts, contactsFilters)
}));

export default mapStateToProps(ContactList);
