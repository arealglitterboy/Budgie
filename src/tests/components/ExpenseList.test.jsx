import React from "react";
import { shallow } from "enzyme";

import { expenses } from "../fixtures/expenses.fixture";
import { ExpenseList } from "../../components/ExpenseList"; // unconnected version of ExpenseList component

// Expenses in expense list
test('should render ExpenseList with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

// No expenses
test('should render empty ExpenseList, empty array', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render empty ExpenseList, no props', () => {
    const wrapper = shallow(<ExpenseList />);
    expect(wrapper).toMatchSnapshot();
});