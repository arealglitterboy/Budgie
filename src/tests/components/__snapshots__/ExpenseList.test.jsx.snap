// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<section
  className="expense-list"
>
  <ExpenseListItem
    amount={1050}
    currency="EUR"
    date={2021-04-09T00:00:00.000Z}
    description="Food"
    id={1}
    key="1"
    note="Bought food for mastication"
  />
  <ExpenseListItem
    amount={50653}
    currency="EUR"
    date={2021-10-12T00:00:00.000Z}
    description="Rent"
    id={2}
    key="2"
    note="Alex (landlord), wanted next month early"
  />
  <ExpenseListItem
    amount={4500}
    currency="MXN"
    date={2010-07-16T00:00:00.000Z}
    description="Credit Card"
    id={3}
    key="3"
    note="Paying off loan to completion"
  />
</section>
`;

exports[`should render empty ExpenseList, empty array 1`] = `
<section
  className="expense-list"
>
  <p
    className="expense-list__empty-list"
  >
    Start by adding an expense
  </p>
</section>
`;

exports[`should render empty ExpenseList, no props (except today, as this sets the values of the expense list items) 1`] = `
<section
  className="expense-list"
>
  <p
    className="expense-list__empty-list"
  >
    Start by adding an expense
  </p>
</section>
`;

exports[`should render empty ExpenseList, no props 1`] = `
<section
  className="expense-list"
>
  <p
    className="expense-list__empty-list"
  >
    Start by adding an expense
  </p>
</section>
`;
